pisa/core/translation.py

```diff
@@ -360,38 +362,56 @@ def lookup(sample, flat_hist, binning):
 
 
 @myjit
-def find_index(x, bin_edges):
-    """simple binary search
+def find_index(val, bin_edges):
+    """Find index in binning for `val`. If `val` is below binning range or is
+    nan, return -1; if `val` is above binning range, return num_bins. Edge
+    inclusivity/exclusivity is defined as .. ::
+
+        [ bin 0 ) [ bin 1 ) ... [ bin num_bins-1 ]
+
+    Using these indices to produce histograms should yield identical results
+    (ignoring underflow and overflow, which `find_index` has) that are
+    equivalent to those produced by ``numpy.histogramdd``.
+
+    Parameters
+    ----------
+    val : scalar
+        Value for which to find bin index
+
+    bin_edges : 1d numpy ndarray
+        Must be monotonically increasing, and all bins are adjacent
+
+    Returns
+    -------
+    idx : int in [-1, num_bins]
 
-    direct transformations instead of search
     """
-    # TODO: support lin and log binnings with
-
-    #
-    # First check: ouside binning
-    #
-    if x < bin_edges[0]:
-        return -1
-    elif x > bin_edges[-1]:
-        return len(bin_edges) - 1
+    # TODO: support fast computation for lin and log binnings?
+
+    num_bins = len(bin_edges) - 1
+    underflow_idx = -1
+    overflow_idx = num_bins
+
+    # First check: NaN or ouside binning?
+    if np.isnan(val):
+        idx = underflow_idx
+    elif val < bin_edges[0]:
+        idx = underflow_idx
+    elif val > bin_edges[-1]:
+        idx = overflow_idx
     else:
-        #
-        # Now handle middle cases
-        #
-        first = 0
-        last = len(bin_edges) - 1
-        while first <= last:
-            i = int((first + last)/2)
-            if x > bin_edges[i]:
-
-                if x <= bin_edges[i+1]:
-                    break
-                else:
-                    first = i + 1
+        # Binary search within binning (inclusive of left and right edges)
+        left_idx = 0
+        right_idx = num_bins
+        while left_idx < right_idx:
+            idx = left_idx + ((right_idx - left_idx) >> 1)
+            if val >= bin_edges[idx]:
+                left_idx = idx + 1
```

Comment:
i'm not mentally capable of being 100% sure this is correct, but it looks like. Also I see you tested it.

Reply:
I updated some variable names and added comments in hopes that the logic is a bit easier to follow (binary search still isn't trivial, but should be able to reason it out). And added more unit tests (e.g. for single-bin binnings).